compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -175,10 +175,16 @@ object Checking {
         if (cls.isOneOf(AbstractOrTrait))
           ctx.error(CantInstantiateAbstractClassOrTrait(cls, isTrait = cls.is(Trait)), posd.sourcePos)
         if (!cls.is(Module)) {
+          def (tp: Type).stripOpaques: Type = tp match
+            case RefinedType(parent, name, _) if cls.info.decl(name).symbol.isOpaqueAlias =>
+              parent.stripOpaques
+            case _ =>
+              tp
           // Create a synthetic singleton type instance, and check whether
           // it conforms to the self type of the class as seen from that instance.
           val stp = SkolemType(tp)
-          val selfType = cls.asClass.givenSelfType.asSeenFrom(stp, cls)
+          val selfType =
+            cls.asClass.givenSelfType.stripOpaques.asSeenFrom(stp, cls)
```

Comment:
Maybe stripOpaque should be a boolean parameter of `givenSelfTypes` ?

Reply:
I think it's safer to keep givenSelfType as is.  It's really just when checking that we need to strip the refinements.